File: frontend/lib/estimator.ts
export type DistributionModel = 'pre_packed' | 'list' | 'client_choice';
export type StockBand = 'plenty' | 'low' | 'out';

export const PLENTY_THRESHOLD = 20;
export const LOW_THRESHOLD = 5;

export const DEFAULT_LBS_PER_PERSON: Record<string, number> = {
  produce: 2.0,
  protein: 1.5,
  dairy: 1.0,
  grains: 1.5,
  diapers: 0.5,
  hygiene: 0.3,
  'canned goods': 2.0,
  'halal items': 1.5,
  halal: 1.5,
};

export const BAND_SNAP_QTY: Record<StockBand, number> = {
  plenty: 25,
  low: 12,
  out: 0,
};

const PREDICTION_CONFIDENCE_AFTER_TRUTH = 0.8;
const PREDICTION_CONFIDENCE_DECAY = 0.9;
const PREDICTION_CONFIDENCE_FLOOR = 0.4;

export interface ShelfSnapshot {
  category_id?: number;
  category_name: string;
  category_emoji?: string;
  estimated_qty?: number | null;
  band?: StockBand | string | null;
  lbs_per_person?: number | null;
  source?: string | null;
  confidence?: number | null;
  minutes_ago?: number | null;
}

export interface OrderItem {
  category_id?: number;
  category_name?: string;
  quantity: number;
}

export interface Depletion {
  category_id?: number;
  category_name: string;
  category_emoji: string;
  depleted: number;
  previous_qty: number;
  remaining_qty: number;
  band: StockBand;
  previous_band: string;
  confidence: number;
  source: 'prediction';
  lbs_per_person: number;
}

export function qtyToBand(qty: number | null | undefined): StockBand {
  const remaining = qty == null ? 0 : Number(qty);
  if (remaining > PLENTY_THRESHOLD) return 'plenty';
  if (remaining >= LOW_THRESHOLD) return 'low';
  return 'out';
}

export function bandToQty(band: string, currentQty?: number | null): number {
  const normalized = (band || 'low').toLowerCase() as StockBand;
  const current = currentQty == null ? null : Number(currentQty);
  if (normalized === 'plenty') {
    if (current != null && current > PLENTY_THRESHOLD) return current;
    return BAND_SNAP_QTY.plenty;
  }
  if (normalized === 'low') {
    if (current != null && current >= LOW_THRESHOLD && current <= PLENTY_THRESHOLD) return current;
    return BAND_SNAP_QTY.low;
  }
  if (current != null && current < LOW_THRESHOLD) return Math.max(0, current);
  return BAND_SNAP_QTY.out;
}

export function nextPredictionConfidence(
  previousConfidence?: number | null,
  previousSource?: string | null
): number {
  const source = (previousSource || '').toLowerCase();
  if (['volunteer_correction', 'intake_photo', 'manual', 'client_feedback'].includes(source)) {
    return PREDICTION_CONFIDENCE_AFTER_TRUTH;
  }
  const prev = previousConfidence == null ? PREDICTION_CONFIDENCE_AFTER_TRUTH : Number(previousConfidence);
  return Math.round(Math.max(PREDICTION_CONFIDENCE_FLOOR, prev * PREDICTION_CONFIDENCE_DECAY) * 100) / 100;
}

export function normalizeCategoryKey(name: string): string {
  return (name || '').trim().toLowerCase();
}

export function lbsPerPersonFor(categoryName: string, explicit?: number | null): number {
  if (explicit != null) return Number(explicit);
  const key = normalizeCategoryKey(categoryName);
  if (key in DEFAULT_LBS_PER_PERSON) return DEFAULT_LBS_PER_PERSON[key];
  for (const [needle, rate] of Object.entries(DEFAULT_LBS_PER_PERSON)) {
    if (key.includes(needle)) return rate;
  }
  return 1;
}

export function depletionForCategory(opts: {
  distributionModel: string;
  householdSize: number;
  lbsPerPerson: number;
  orderedQty?: number | null;
}): number {
  const model = (opts.distributionModel || 'client_choice').toLowerCase();
  const size = Math.max(1, Number(opts.householdSize) || 1);
  const rate = Number(opts.lbsPerPerson);
  if (model === 'pre_packed') return Math.round(rate * 100) / 100;
  if (model === 'list' && opts.orderedQty != null) {
    return Math.round(Math.max(0, Number(opts.orderedQty)) * 100) / 100;
  }
  return Math.round(size * rate * 100) / 100;
}

export function applyDepletion(currentQty: number | null | undefined, depleted: number): number {
  const remaining = currentQty == null ? 0 : Number(currentQty);
  return Math.round(Math.max(0, remaining - Math.max(0, depleted)) * 100) / 100;
}

export function qtyFromBand(band?: string | null, estimatedQty?: number | null): number {
  if (estimatedQty != null) return Number(estimatedQty);
  const key = (band || 'low').toLowerCase() as StockBand;
  return BAND_SNAP_QTY[key] ?? 12;
}

export function estimateCheckin(opts: {
  distributionModel: string;
  householdSize: number;
  shelfRows: ShelfSnapshot[];
  orderItems?: OrderItem[] | null;
}): {
  distribution_model: DistributionModel;
  household_size: number;
  estimated_lbs: number;
  depletions: Depletion[];
} {
  const ordersById = new Map<number, number>();
  const ordersByName = new Map<string, number>();
  for (const item of opts.orderItems || []) {
    if (item.quantity == null) continue;
    if (item.category_id != null) ordersById.set(item.category_id, Number(item.quantity));
    if (item.category_name) ordersByName.set(normalizeCategoryKey(item.category_name), Number(item.quantity));
  }

  const hasExplicitOrder = ordersById.size > 0 || ordersByName.size > 0;
  const model = (opts.distributionModel || 'client_choice').toLowerCase() as DistributionModel;
  const depletions: Depletion[] = [];
  let estimatedLbs = 0;

  for (const row of opts.shelfRows) {
    const name = row.category_name || '';
    const rate = lbsPerPersonFor(name, row.lbs_per_person);
    let ordered: number | null = null;
    if (row.category_id != null && ordersById.has(row.category_id)) {
      ordered = ordersById.get(row.category_id)!;
    } else if (ordersByName.has(normalizeCategoryKey(name))) {
      ordered = ordersByName.get(normalizeCategoryKey(name))!;
    }

    let depleted = depletionForCategory({
      distributionModel: model,
      householdSize: opts.householdSize,
      lbsPerPerson: rate,
      orderedQty: ordered,
    });
    if (model === 'list' && hasExplicitOrder && ordered == null) {
      depleted = 0;
    }

    const currentQty = qtyFromBand(row.band, row.estimated_qty);
    const remaining = applyDepletion(currentQty, depleted);
    const band = qtyToBand(remaining);
    estimatedLbs += depleted;
    depletions.push({
      category_id: row.category_id,
      category_name: name,
      category_emoji: row.category_emoji || '📦',
      depleted,
      previous_qty: currentQty,
      remaining_qty: remaining,
      band,
      previous_band: String(row.band || qtyToBand(currentQty)),
      confidence: nextPredictionConfidence(row.confidence, row.source),
      source: 'prediction',
      lbs_per_person: rate,
    });
  }

  return {
    distribution_model: model,
    household_size: Math.max(1, Number(opts.householdSize) || 1),
    estimated_lbs: Math.round(estimatedLbs * 100) / 100,
    depletions,
  };
}

export function snapCorrections(
  corrections: Array<{ category_id?: number; category_name?: string; band: string; estimated_qty?: number | null; category_emoji?: string }>,
  shelfRows: ShelfSnapshot[]
) {
  const byId = new Map<number, ShelfSnapshot>();
  const byName = new Map<string, ShelfSnapshot>();
  for (const row of shelfRows) {
    if (row.category_id != null) byId.set(row.category_id, row);
    byName.set(normalizeCategoryKey(row.category_name), row);
  }

  return corrections.map((correction) => {
    const row =
      (correction.category_id != null ? byId.get(correction.category_id) : undefined) ||
      byName.get(normalizeCategoryKey(correction.category_name || '')) ||
      {};
    const band = (correction.band || (row as ShelfSnapshot).band || 'low').toLowerCase() as StockBand;
    const currentQty = correction.estimated_qty ?? (row as ShelfSnapshot).estimated_qty;
    return {
      category_id: correction.category_id ?? (row as ShelfSnapshot).category_id,
      category_name: correction.category_name || (row as ShelfSnapshot).category_name,
      category_emoji: correction.category_emoji || (row as ShelfSnapshot).category_emoji || '📦',
      band,
      estimated_qty: bandToQty(band, currentQty),
      source: 'volunteer_correction' as const,
      confidence: 1.0,
    };
  });
}

export function defaultPackLbs(shelfRows: ShelfSnapshot[]): number {
  const total = shelfRows.reduce((sum, row) => sum + lbsPerPersonFor(row.category_name, row.lbs_per_person), 0);
  return Math.round(total * 100) / 100 || 10.3;
}

File: backend/estimator.py
"""Predict-and-correct inventory state estimator.

Band mapping (lbs remaining):
    qty > 20          → plenty (green)
    5 <= qty <= 20    → low (amber)
    qty < 5           → out (red)

Depletion depends on the pantry's distribution_model:
    client_choice  household_size × lbs_per_person  (shop-the-shelves estimate)
    pre_packed     1 fixed pack per check-in (lbs_per_person per category)
    list           explicit order quantities, else typical list = client_choice math
"""

from __future__ import annotations

from typing import Any, Iterable, Mapping, Optional, Sequence

PLENTY_THRESHOLD = 20.0  # exclusive: qty > 20 is plenty
LOW_THRESHOLD = 5.0  # inclusive lower bound of "low"

DEFAULT_LBS_PER_PERSON: dict[str, float] = {
    "produce": 2.0,
    "protein": 1.5,
    "dairy": 1.0,
    "grains": 1.5,
    "diapers": 0.5,
    "hygiene": 0.3,
    "canned goods": 2.0,
    "halal items": 1.5,
    "halal": 1.5,
}

BAND_SNAP_QTY = {
    "plenty": 25.0,
    "low": 12.0,
    "out": 0.0,
}

PREDICTION_CONFIDENCE_AFTER_TRUTH = 0.80
PREDICTION_CONFIDENCE_DECAY = 0.90
PREDICTION_CONFIDENCE_FLOOR = 0.40


def qty_to_band(qty: Optional[float]) -> str:
    """Map remaining estimated quantity onto a stock band."""
    remaining = 0.0 if qty is None else float(qty)
    if remaining > PLENTY_THRESHOLD:
        return "plenty"
    if remaining >= LOW_THRESHOLD:
        return "low"
    return "out"


def band_to_qty(band: str, current_qty: Optional[float] = None) -> float:
    """Snap a volunteer-confirmed band to a ground-truth quantity.

    If the current estimate already sits inside the confirmed band, keep it.
    Otherwise jump to a representative quantity for that band.
    """
    normalized = (band or "low").lower()
    current = None if current_qty is None else float(current_qty)

    if normalized == "plenty":
        if current is not None and current > PLENTY_THRESHOLD:
            return current
        return BAND_SNAP_QTY["plenty"]
    if normalized == "low":
        if current is not None and LOW_THRESHOLD <= current <= PLENTY_THRESHOLD:
            return current
        return BAND_SNAP_QTY["low"]
    if current is not None and current < LOW_THRESHOLD:
        return max(0.0, current)
    return BAND_SNAP_QTY["out"]


def next_prediction_confidence(previous_confidence: Optional[float], previous_source: Optional[str]) -> float:
    """Decay confidence after each predicted depletion; reset higher after a human/photo update."""
    source = (previous_source or "").lower()
    if source in {"volunteer_correction", "intake_photo", "manual", "client_feedback"}:
        return PREDICTION_CONFIDENCE_AFTER_TRUTH
    prev = PREDICTION_CONFIDENCE_AFTER_TRUTH if previous_confidence is None else float(previous_confidence)
    return round(max(PREDICTION_CONFIDENCE_FLOOR, prev * PREDICTION_CONFIDENCE_DECAY), 2)


def normalize_category_key(name: str) -> str:
    return (name or "").strip().lower()


def lbs_per_person_for(category_name: str, explicit: Optional[float] = None) -> float:
    if explicit is not None:
        return float(explicit)
    key = normalize_category_key(category_name)
    if key in DEFAULT_LBS_PER_PERSON:
        return DEFAULT_LBS_PER_PERSON[key]
    for needle, rate in DEFAULT_LBS_PER_PERSON.items():
        if needle in key:
            return rate
    return 1.0


def depletion_for_category(
    *,
    distribution_model: str,
    household_size: int,
    lbs_per_person: float,
    ordered_qty: Optional[float] = None,
) -> float:
    """How many units leave the shelf for one category on this check-in."""
    model = (distribution_model or "client_choice").lower()
    size = max(1, int(household_size))
    rate = float(lbs_per_person)

    if model == "pre_packed":
        return round(rate, 2)
    if model == "list":
        if ordered_qty is not None:
            return round(max(0.0, float(ordered_qty)), 2)
        return round(size * rate, 2)
    # client_choice / shop the shelves
    return round(size * rate, 2)


def apply_depletion(
    current_qty: Optional[float],
    depleted: float,
) -> float:
    remaining = 0.0 if current_qty is None else float(current_qty)
    return round(max(0.0, remaining - max(0.0, float(depleted))), 2)


def qty_from_band(band: Optional[str], estimated_qty: Optional[float] = None) -> float:
    """Seed a numeric estimate when the latest shelf row has a band but no quantity."""
    if estimated_qty is not None:
        return float(estimated_qty)
    return float(BAND_SNAP_QTY.get((band or "low").lower(), 12.0))


def estimate_checkin(
    *,
    distribution_model: str,
    household_size: int,
    shelf_rows: Sequence[Mapping[str, Any]],
    order_items: Optional[Iterable[Mapping[str, Any]]] = None,
) -> dict[str, Any]:
    """Pure predict-and-correct step for one household check-in.

    shelf_rows items support keys:
        category_id, category_name, category_emoji, estimated_qty, band,
        lbs_per_person, source, confidence
    order_items items support keys:
        category_id, category_name, quantity
    """
    orders_by_id: dict[Any, float] = {}
    orders_by_name: dict[str, float] = {}
    for item in order_items or []:
        qty = item.get("quantity")
        if qty is None:
            continue
        if item.get("category_id") is not None:
            orders_by_id[item["category_id"]] = float(qty)
        if item.get("category_name"):
            orders_by_name[normalize_category_key(str(item["category_name"]))] = float(qty)

    depletions: list[dict[str, Any]] = []
    estimated_lbs = 0.0

    for row in shelf_rows:
        name = row.get("category_name") or ""
        cat_id = row.get("category_id")
        rate = lbs_per_person_for(name, row.get("lbs_per_person"))
        ordered = None
        if cat_id in orders_by_id:
            ordered = orders_by_id[cat_id]
        elif normalize_category_key(name) in orders_by_name:
            ordered = orders_by_name[normalize_category_key(name)]

        depleted = depletion_for_category(
            distribution_model=distribution_model,
            household_size=household_size,
            lbs_per_person=rate,
            ordered_qty=ordered,
        )
        # List model with an explicit order: skip categories they didn't take.
        if (distribution_model or "").lower() == "list" and (orders_by_id or orders_by_name) and ordered is None:
            depleted = 0.0

        current_qty = qty_from_band(row.get("band"), row.get("estimated_qty"))
        remaining = apply_depletion(current_qty, depleted)
        band = qty_to_band(remaining)
        confidence = next_prediction_confidence(row.get("confidence"), row.get("source"))

        estimated_lbs += depleted
        depletions.append(
            {
                "category_id": cat_id,
                "category_name": name,
                "category_emoji": row.get("category_emoji") or "📦",
                "depleted": depleted,
                "previous_qty": current_qty,
                "remaining_qty": remaining,
                "band": band,
                "previous_band": row.get("band") or qty_to_band(current_qty),
                "confidence": confidence,
                "source": "prediction",
                "lbs_per_person": rate,
            }
        )

    return {
        "distribution_model": (distribution_model or "client_choice").lower(),
        "household_size": max(1, int(household_size)),
        "estimated_lbs": round(estimated_lbs, 2),
        "depletions": depletions,
    }


def snap_corrections(
    corrections: Sequence[Mapping[str, Any]],
    shelf_rows: Sequence[Mapping[str, Any]],
) -> list[dict[str, Any]]:
    """Turn volunteer closing-check bands into ground-truth shelf rows (confidence 1.0)."""
    by_id = {row.get("category_id"): row for row in shelf_rows if row.get("category_id") is not None}
    by_name = {normalize_category_key(str(row.get("category_name") or "")): row for row in shelf_rows}

    snapped: list[dict[str, Any]] = []
    for correction in corrections:
        cat_id = correction.get("category_id")
        name = correction.get("category_name") or ""
        row = by_id.get(cat_id) if cat_id is not None else None
        if row is None:
            row = by_name.get(normalize_category_key(name), {})
        band = (correction.get("band") or row.get("band") or "low").lower()
        current_qty = correction.get("estimated_qty")
        if current_qty is None:
            current_qty = row.get("estimated_qty")
        qty = band_to_qty(band, current_qty)
        snapped.append(
            {
                "category_id": cat_id if cat_id is not None else row.get("category_id"),
                "category_name": name or row.get("category_name"),
                "category_emoji": correction.get("category_emoji") or row.get("category_emoji") or "📦",
                "band": band,
                "estimated_qty": qty,
                "source": "volunteer_correction",
                "confidence": 1.0,
            }
        )
    return snapped
